src/main/java/ml/shifu/shifu/core/dtrain/wdl/WDLParallelGradient.java

```diff
@@ -142,22 +140,31 @@ private void adjustTestSet(int start, int miniBatchSize) {
         }
     }
 
+    public WDLParams doCompute() {
+        return doCompute(false);
+    }
+
     /**
      * Batch training in parallel with {@link WDLParallelGradient#threadNumber} threads
-     * 
+     *
+     * @param shuffle whether to shuffle data before train
      * @return
      *      the combined gradients result
      */
-    public WDLParams doCompute() {
+    public WDLParams doCompute(boolean shuffle) {
         long start = System.currentTimeMillis();
+        Shuffler shuffler = null;
+        if (shuffle) {
+            shuffler = new PermutationShuffler(this.trainData.size());
```

Comment:
Each iteration need to create one shuffler instance?

Reply:
I did some refactoring. Now each worker holds one Shuffler instance, and did a in-place refresh before each iteration.